transport/http/do_test.go

```diff
@@ -0,0 +1,369 @@
+			method:  "POST",
+			timeout: 0,
+		},
+	}
+
+	httpTransport := NewTransport()
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			server := httptest.NewServer(http.HandlerFunc(
+				func(w http.ResponseWriter, r *http.Request) {
+					defer r.Body.Close()
+				},
+			))
+			out := httpTransport.NewSingleOutbound(server.URL)
+
+			require.NoError(t, out.Start(), "failed to start outbound")
+			defer out.Stop()
+
+			ctx := context.Background()
+			ctx, cancel := context.WithTimeout(ctx, tt.timeout)
+			defer cancel()
+
+			req, err := http.NewRequest(tt.method, server.URL, bytes.NewReader([]byte("world")))
+			require.NoError(t, err)
+			req = req.WithContext(ctx)
+
+			client := http.Client{Transport: out}
+
+			res, err := client.Do(req)
+			assert.Equal(t, "context deadline exceeded", ctx.Err().Error())
+			assert.Error(t, err)
+			assert.Nil(t, res)
+		})
+	}
+}
+
+func TestHTTPApplicationError(t *testing.T) {
+	tests := []struct {
+		desc   string
+		status string
+	}{
+		{
+			desc:   "ok",
+			status: "success",
+		},
+		{
+			desc:   "error",
+			status: "error",
+		},
+		{
+			desc:   "not an error",
+			status: "lolwut",
+		},
+	}
+
+	httpTransport := NewTransport()
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			server := httptest.NewServer(http.HandlerFunc(
+				func(w http.ResponseWriter, r *http.Request) {
+					w.Header().Add("Status", tt.status)
+					defer r.Body.Close()
+				},
+			))
+			defer server.Close()
+
+			out := httpTransport.NewSingleOutbound(server.URL)
+
+			require.NoError(t, out.Start(), "failed to start outbound")
+			defer out.Stop()
+
+			ctx := context.Background()
+			ctx, cancel := context.WithTimeout(ctx, 100*testtime.Millisecond)
+			defer cancel()
+
+			req, err := http.NewRequest("GET", server.URL, bytes.NewReader([]byte("world")))
+			require.NoError(t, err)
+			req = req.WithContext(ctx)
+
+			client := http.Client{Transport: out}
+
+			res, err := client.Do(req)
+
+			assert.Equal(t, res.Header.Get("Status"), tt.status, "%v: application status", tt.desc)
+
+			require.NoError(t, err)
+			require.NoError(t, res.Body.Close())
+		})
+	}
+}
+
+func TestHTTPCallFailures(t *testing.T) {
+	notFoundServer := httptest.NewServer(http.NotFoundHandler())
+	defer notFoundServer.Close()
+
+	internalErrorServer := httptest.NewServer(http.HandlerFunc(
+		func(w http.ResponseWriter, r *http.Request) {
+			http.Error(w, "great sadness", http.StatusInternalServerError)
+		}))
+	defer internalErrorServer.Close()
+
+	httpTransport := NewTransport()
+
+	tests := []struct {
+		desc     string
+		url      string
+		messages []string
+	}{
+		{
+			"not a URL",
+			"not a URL",
+			[]string{"protocol scheme"},
+		},
+		{
+			"404",
+			notFoundServer.URL,
+			[]string{"404", "page not found"},
+		},
+		{
+			"failed due to error",
+			internalErrorServer.URL,
+			[]string{"great sadness"},
+		},
+		{
+			"fail for a fully qualified URL",
+			"http://example.com/",
```

Comment:
I’m surprised this case didn’t require any new code! At least, I don’t see it. Is it failing for a reason that is not that the path is fully qualified? Maybe it’s attempting to send a request anyway? Can we get more detail in the error message to lock it down?